Update to canonical 0.6.2 breaks contracts

> **Note**
> We had to yank this version, since it's either a mistake made in the last version, or if it's on purpose and therefore a breaking change – and we should release a major version.

Not sure why this is wasn't catch in the tests, it looks like it's a scope issue, so maybe it's because some unit tests was used instead of integration tests (as rule of thumb we should always use integration tests to avoid this specific scenario unless we're purposely testing internal behaviors):

```
   Compiling canonical v0.6.2
error[E0412]: cannot find type `Id` in this scope
  --> /Users/zer0/.cargo/registry/src/github.com-1ecc6299db9ec823/canonical-0.6.2/src/store/bridge.rs:42:28
   |
42 |     pub fn take_bytes(id: &Id) -> Result<Vec<u8>, CanonError> {
   |                            ^^ not found in this scope
   |
help: consider importing this struct
   |
9  | use crate::Id;
   |

error[E0412]: cannot find type `Vec` in this scope
  --> /Users/zer0/.cargo/registry/src/github.com-1ecc6299db9ec823/canonical-0.6.2/src/store/bridge.rs:42:42
   |
42 |     pub fn take_bytes(id: &Id) -> Result<Vec<u8>, CanonError> {
   |                                          ^^^ not found in this scope
   |
help: consider importing one of these items
   |
9  | use alloc::vec::Vec;
   |
9  | use crate::implementations::alloc_impls::Vec;
   |

error[E0433]: failed to resolve: use of undeclared type `Vec`
  --> /Users/zer0/.cargo/registry/src/github.com-1ecc6299db9ec823/canonical-0.6.2/src/store/bridge.rs:45:23
   |
45 |         let mut buf = Vec::with_capacity(len);
   |                       ^^^ not found in this scope
   |
help: consider importing one of these items
   |
9  | use alloc::vec::Vec;
   |
9  | use crate::implementations::alloc_impls::Vec;
   |

error[E0425]: cannot find value `vec` in this scope
  --> /Users/zer0/.cargo/registry/src/github.com-1ecc6299db9ec823/canonical-0.6.2/src/store/bridge.rs:48:12
   |
48 |         Ok(vec)
   |            ^^^ not found in this scope

error[E0412]: cannot find type `Vec` in this scope
  --> /Users/zer0/.cargo/registry/src/github.com-1ecc6299db9ec823/canonical-0.6.2/src/store/mod.rs:45:49
   |
45 |     pub(crate) fn take_bytes(id: &Id) -> Result<Vec<u8>, CanonError> {
   |                                                 ^^^ not found in this scope
   |
help: consider importing one of these items
   |
7  | use alloc::vec::Vec;
   |
7  | use crate::implementations::alloc_impls::Vec;
   |

error: aborting due to 5 previous errors

Some errors have detailed explanations: E0412, E0425, E0433.
For more information about an error, try `rustc --explain E0412`.
error: could not compile `canonical`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
make[1]: *** [all] Error 101
make: *** [tests/contracts/block_height/.] Error 2
```

--- a/canon/CHANGELOG.md
+++ b/canon/CHANGELOG.md
@@ -2,6 +2,15 @@
 
 ## [Unreleased]
 
+## [0.6.3] 2021-05-26
+
+### Added
+- Add varint support for `u128` and `i128` by converting them to two `u64`s
+- Add encoding/decoding tests for all integer types
+
+### Changed
+- Change library `integer-encoding` to `dusk-varint`
+
 ## [0.6.2] 2021-05-20
 
 ### Added
@@ -115,7 +124,8 @@
 
 Initial release
 
-[Unreleased]: https://github.com/dusk-network/canonical/compare/canonical-0.6.2...HEAD
+[Unreleased]: https://github.com/dusk-network/canonical/compare/canonical-0.6.3...HEAD
+[0.6.3]: https://github.com/dusk-network/canonical/compare/canonical-0.6.2...canonical-0.6.3
 [0.6.2]: https://github.com/dusk-network/canonical/compare/canonical-0.6.1...canonical-0.6.2
 [0.6.1]: https://github.com/dusk-network/canonical/compare/canonical-0.6.0...canonical-0.6.1
 [0.6.0]: https://github.com/dusk-network/canonical/compare/canonical-0.5.3...canonical-0.6.0

--- a/canon/Cargo.toml
+++ b/canon/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "canonical"
-version = "0.6.2"
+version = "0.6.3"
 authors = ["Kristoffer Ström <kristoffer@dusk.network>"]
 edition = "2018"
 readme = "../README.md"
@@ -13,7 +13,7 @@ license = "MPL-2.0"
 blake2b_simd = { version = "0.3", default-features = false }
 cfg-if = "1.0.0"
 array-init = "2.0"
-integer-encoding = "3.0"
+dusk-varint = "0.1"
 
 [dev-dependencies]
 canonical_derive = { path = "../canon_derive", version = "0.6" }

--- a/canon/src/implementations.rs
+++ b/canon/src/implementations.rs
@@ -7,35 +7,25 @@
 #![allow(clippy::empty_loop)]
 use core::marker::PhantomData;
 use core::mem;
-use integer_encoding::VarInt;
+use dusk_varint::VarInt;
 
 use crate::{Canon, CanonError, Sink, Source};
 
-macro_rules! number {
-    ($number:ty, $size:expr) => {
-        impl Canon for $number {
-            fn encode(&self, sink: &mut Sink) {
-                sink.copy_bytes(&self.to_be_bytes())
-            }
+impl Canon for u8 {
+    fn encode(&self, sink: &mut Sink) {
+        sink.copy_bytes(&self.to_be_bytes())
+    }
 
-            fn decode(source: &mut Source) -> Result<Self, $crate::CanonError> {
-                let mut bytes = [0u8; $size];
-                bytes.copy_from_slice(source.read_bytes($size));
-                Ok(<$number>::from_be_bytes(bytes))
-            }
+    fn decode(source: &mut Source) -> Result<Self, CanonError> {
+        let mut bytes = [0u8; 1];
+        bytes.copy_from_slice(source.read_bytes(1));
+        Ok(u8::from_be_bytes(bytes))
+    }
 
-            fn encoded_len(&self) -> usize {
-                $size
-            }
-        }
-    };
+    fn encoded_len(&self) -> usize {
+        1
+    }
 }
-
-number!(u8, 1);
-number!(i8, 1);
-
-number!(u128, 16);
-number!(i128, 16);
 
 macro_rules! varint {
     ($varint:ty) => {
@@ -80,6 +70,53 @@ varint!(i32);
 
 varint!(u64);
 varint!(i64);
+
+impl Canon for u128 {
+    fn encode(&self, sink: &mut Sink) {
+        let high: u64 = (self >> 64) as u64;
+        let low: u64 = *self as u64;
+
+        high.encode(sink);
+        low.encode(sink);
+    }
+
+    fn decode(source: &mut Source) -> Result<Self, CanonError> {
+        let high = u64::decode(source)?;
+        let low = u64::decode(source)?;
+
+        Ok((low as u128) + ((high as u128) << 64))
+    }
+
+    fn encoded_len(&self) -> usize {
+        let high: u64 = (self >> 64) as u64;
+        let low: u64 = *self as u64;
+        high.encoded_len() + low.encoded_len()
+    }
+}
+
+#[inline]
+fn zigzag_encode(from: i128) -> u128 {
+    ((from << 1) ^ (from >> 127)) as u128
+}
+
+#[inline]
+fn zigzag_decode(from: u128) -> i128 {
+    ((from >> 1) ^ (-((from & 1) as i128)) as u128) as i128
+}
+
+impl Canon for i128 {
+    fn encode(&self, sink: &mut Sink) {
+        zigzag_encode(*self).encode(sink)
+    }
+
+    fn decode(source: &mut Source) -> Result<Self, CanonError> {
+        Ok(zigzag_decode(u128::decode(source)?))
+    }
+
+    fn encoded_len(&self) -> usize {
+        zigzag_encode(*self).encoded_len()
+    }
+}
 
 impl Canon for bool {
     fn encode(&self, sink: &mut Sink) {

--- a/canon/src/store/bridge.rs
+++ b/canon/src/store/bridge.rs
@@ -7,7 +7,8 @@
 extern crate alloc;
 
 use crate::canon::CanonError;
-use crate::id::IdHash;
+use crate::id::{Id, IdHash};
+use alloc::vec::Vec;
 
 /// Store usable across ffi-boundraries
 #[derive(Clone, Copy, Default, Debug)]
@@ -45,7 +46,7 @@ impl BridgeStore {
         let mut buf = Vec::with_capacity(len);
         buf.resize_with(len, || 0);
         Self::get(&id.hash(), &mut buf[..])?;
-        Ok(vec)
+        Ok(buf)
     }
 }
 

--- a/canon/src/store/mod.rs
+++ b/canon/src/store/mod.rs
@@ -10,6 +10,7 @@ use core::fmt;
 
 use crate::id::{Id, IdHash};
 use crate::CanonError;
+use alloc::vec::Vec;
 
 cfg_if! {
     if #[cfg(target_arch = "wasm32")] {

--- a/canon/tests/integers.rs
+++ b/canon/tests/integers.rs
@@ -1,0 +1,123 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use canonical::{Canon, Sink, Source};
+
+#[test]
+fn test_u8() {
+    let mut buf = [0u8; 1];
+
+    for i in 0..u8::MAX {
+        let mut sink = Sink::new(&mut buf);
+        i.encode(&mut sink);
+        drop(sink);
+        let mut source = Source::new(&buf);
+        assert_eq!(i, u8::decode(&mut source).unwrap());
+    }
+}
+
+#[test]
+fn test_u16() {
+    let mut buf = [0u8; 3];
+
+    for i in 0..u16::MAX {
+        let mut sink = Sink::new(&mut buf);
+        i.encode(&mut sink);
+        drop(sink);
+        let mut source = Source::new(&buf);
+        assert_eq!(i, u16::decode(&mut source).unwrap());
+    }
+}
+
+#[test]
+fn test_u32() {
+    let mut buf = [0u8; 5];
+
+    let mut i = 0;
+    const STRIDE: u32 = u32::MAX / 1024;
+
+    while i < u32::MAX {
+        i = i.saturating_add(STRIDE);
+
+        let mut sink = Sink::new(&mut buf);
+        i.encode(&mut sink);
+        drop(sink);
+        let mut source = Source::new(&buf);
+        assert_eq!(i, u32::decode(&mut source).unwrap());
+    }
+}
+
+#[test]
+fn test_u64() {
+    let mut buf = [0u8; 10];
+
+    let mut i = 0;
+    const STRIDE: u64 = u64::MAX / 1024;
+
+    while i < u64::MAX {
+        i = i.saturating_add(STRIDE);
+
+        let mut sink = Sink::new(&mut buf);
+        i.encode(&mut sink);
+        drop(sink);
+        let mut source = Source::new(&buf);
+        assert_eq!(i, u64::decode(&mut source).unwrap());
+    }
+}
+
+#[test]
+fn test_u128() {
+    let mut buf = [0u8; 20];
+
+    let mut i = 0;
+    const STRIDE: u128 = u128::MAX / 1024;
+
+    while i < u128::MAX {
+        i = i.saturating_add(STRIDE);
+
+        let mut sink = Sink::new(&mut buf);
+        i.encode(&mut sink);
+        drop(sink);
+        let mut source = Source::new(&buf);
+        assert_eq!(i, u128::decode(&mut source).unwrap());
+    }
+}
+
+#[test]
+fn test_i64() {
+    let mut buf = [0u8; 10];
+
+    let mut i = i64::MIN;
+    const STRIDE: i64 = i64::MAX / 1024;
+
+    while i < i64::MAX {
+        i = i.saturating_add(STRIDE);
+
+        let mut sink = Sink::new(&mut buf);
+        i.encode(&mut sink);
+        drop(sink);
+        let mut source = Source::new(&buf);
+        assert_eq!(i, i64::decode(&mut source).unwrap());
+    }
+}
+
+#[test]
+fn test_i128() {
+    let mut buf = [0u8; 20];
+
+    let mut i = i128::MIN;
+    const STRIDE: i128 = i128::MAX / 1024;
+
+    while i < i128::MAX {
+        i = i.saturating_add(STRIDE);
+
+        let mut sink = Sink::new(&mut buf);
+        i.encode(&mut sink);
+        drop(sink);
+        let mut source = Source::new(&buf);
+        assert_eq!(i, i128::decode(&mut source).unwrap());
+    }
+}

